packages/react-chat/src/contexts/RuntimeContext/useRuntimeState.ts

```diff
@@ -0,0 +1,174 @@
+    setSession((prev) => ({ ...prev, turns: action(prev.turns) }));
+  };
+
+  const addTurn = (turn: TurnProps) => setTurns((prev) => [...prev, turn]);
+
+  const reset = () => setTurns(() => []);
+
+  const interact: SendMessage = async (action: BaseRequest.BaseRequest, message?: string) => {
+    if (sessionRef.current.status === SessionStatus.ENDED) return;
+
+    // create a transcript on the first turn, do this async
+    if (sessionRef.current.turns.length === 1) runtime.saveTranscript();
+
+    handleActions(action);
+
+    const userMessage = message || (isTextRequest(action) ? action.payload : null);
+    if (userMessage) {
+      addTurn({
+        id: cuid(),
+        type: TurnType.USER,
+        message: userMessage,
+        timestamp: Date.now(),
+      });
+    }
+
+    setIndicator(true);
+    const context = await runtime.interact(action);
+    setIndicator(false);
+
+    addTurn({
+      id: cuid(),
+      type: TurnType.SYSTEM,
+      timestamp: Date.now(),
+      ...context,
+    });
+
+    saveSession(assistant.persistence, config.verify.projectID, sessionRef.current);
+
+    const finishedAnimatingAt = context.messages.reduce((acc, message) => {
+      return acc + (message.delay ?? 1000);
+    }, Date.now());
+
+    setLastInteractionAt(finishedAnimatingAt);
+  };
+
+  const launch = async ({ event }: LaunchOptions = {}): Promise<void> => {
+    if (sessionRef.current.turns.length) reset();
+
+    setStatus(SessionStatus.ACTIVE);
+    await interact(event ?? { type: BaseRequest.RequestType.LAUNCH, payload: null });
+  };
+
+  const reply = async (message: string): Promise<void> => interact({ type: BaseRequest.RequestType.TEXT, payload: message });
+
+  // TODO: refactor no-reply system
+  useEffect(() => {
+    let noReplyTimer: NodeJS.Timeout | undefined;
+
+    const checkNoReply = () => {
+      const ready = isStatus(SessionStatus.ACTIVE);
+      if (ready && noReplyTimeout && lastInteractionAt) {
+        const timeSinceLastInteraction = Date.now() - lastInteractionAt;
+        if (timeSinceLastInteraction > noReplyTimeout) {
+          // Trigger no reply action
+          interact({ type: BaseRequest.RequestType.NO_REPLY, payload: null });
+        }
+      }
+      noReplyTimer = setTimeout(checkNoReply, 1000);
+    };
+
+    checkNoReply();
+
+    return () => {
+      clearTimeout(noReplyTimer);
+    };
+  }, [noReplyTimeout, lastInteractionAt]);
+
+  const open = async () => {
+    setOpen(true);
+
+    if (isStatus(SessionStatus.IDLE)) {
+      await launch();
+    }
+  };
+
+  const close = () => setOpen(false);
+
+  return {
+    state: {
+      session,
+      isOpen,
+      indicator,
+    },
+    api: {
+      launch,
+      reply,
+      open,
+      interact,
+      close,
+      addTurn,
+      feedback: runtime.saveFeedback,
+      setStatus,
+      isStatus,
+      reset,
+
+      // these are meant to be static, so bundling them with the API
+      assistant,
+      config,
```

Comment:
we could add a separate `meta` section or something for static stuff that isn't state